core/src/main/java/org/elasticsearch/index/query/BoolQueryParser.java

```diff
@@ -71,14 +71,12 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
                 switch (currentFieldName) {
                 case "must":
                     query = parseContext.parseInnerQueryBuilder();
-                    if (query != null) {
-                        mustClauses.add(query);
-                    }
+                    mustClauses.add(query);
                     break;
                 case "should":
                     query = parseContext.parseInnerQueryBuilder();
-                    if (query != null) {
-                        shouldClauses.add(query);
+                    shouldClauses.add(query);
+                    if (query != EmptyQueryBuilder.PROTOTYPE) {
                         if (parseContext.isFilter() && minimumShouldMatch == null) {
```

Comment:
The original code checks the lucene query for `null` here and only sets minimumShouldMatch=1 (if not yet specified) for the first should clause added. Since we know that EmtpyQueryBuilder will not add a Should-Clause later we also don't have to set `minimumShouldMatch` yet.